bdk_electrum_example/src/electrum.rs

```diff
@@ -0,0 +1,236 @@
+                .block_header(existing_height as usize)?
+                .block_hash();
+            sparse_chain
+                .insert_checkpoint(BlockId {
+                    height: existing_height,
+                    hash: current_hash,
+                })
+                .expect("This never errors because we are working with a fresh chain");
+
+            if current_hash == existing_hash {
+                break;
+            }
+        }
+
+        // Insert the new tip
+        let (tip_height, tip_hash) = self.get_tip()?;
+        sparse_chain
+            .insert_checkpoint(BlockId {
+                height: tip_height,
+                hash: tip_hash,
+            })
+            .expect("This never errors because we are working with a fresh chain");
+
+        let mut keychain_index_update = BTreeMap::new();
+
+        let mut reorgred_tx = Vec::new();
+
+        // Fetch Keychain's last_active_index and all related txids.
+        // Add them into the KeyChainScan
+        for (keychain, mut scripts) in scripts.into_iter() {
+            let mut last_active_index = 0;
+            let mut unused_script_count = 0usize;
+            let mut script_history_txid = Vec::<(Txid, TxHeight)>::new();
+
+            loop {
+                if let Some((index, script)) = scripts.next() {
+                    let history = self
+                        .script_get_history(&script)?
+                        .iter()
+                        .map(|history_result| {
+                            if history_result.height > 0
+                                && (history_result.height as u32) < tip_height
+                            {
+                                return (
+                                    history_result.tx_hash,
+                                    TxHeight::Confirmed(history_result.height as u32),
+                                );
+                            } else {
+                                return (history_result.tx_hash, TxHeight::Unconfirmed);
+                            };
+                        })
+                        .collect::<Vec<(Txid, TxHeight)>>();
+
+                    if history.is_empty() {
+                        unused_script_count += 1;
+                    } else {
+                        last_active_index = index;
+                        script_history_txid.extend(history.iter());
+                        unused_script_count = 0;
+                    }
+
+                    if unused_script_count >= stop_gap.unwrap_or(usize::MAX) {
+                        break;
+                    }
+                } else {
+                    break;
+                }
+            }
+
+            for (txid, index) in script_history_txid {
+                if let Err(err) = sparse_chain.insert_tx(txid, index) {
+                    match err {
+                        InsertTxErr::TxTooHigh => {
+                            /* Don't care about new transactions confirmed while syncing */
```

Comment:
since you checked that no tx is higher than tip I think this should be `unreachable!(..)`.

Reply:
Makes sense.. Done..